### PR TITLE
[NFCI][SYCL] Refactor reductions implementations

### PR DESCRIPTION
**Pull request by aelovikov-intel**
The purpose of this change is

1) Make it clear that sycl::range version always delegates to some
   sycl::nd_range implementation.

2) Flatten all existing implementations. That required splitting
   reduction_parallel_for_basic_impl depending on
   Reduction::has_fast_reduce. However, I also inlined its helper
   routines (that now had unambigious branches for "if
   constexpr (Reduction::*)") so the ratio between duplicate/unique code
   isn't bad at all.

3) Make a unique dispatching entry and have all the implementations
   provide the same interface. I plan to use it in unit-tests to bypass
   the dispatch and test all the implementation directly (when
   applicable based on nd_range/BinOp/HW/etc.).

I'd say 90% of the change is straightforward code movement with the
exceptions of

  - Inlining described above
  - Simplifying __sycl_reduction_kernel helper enabled by this change
  - Factoring out forward declarations in handler.hpp into a separate
    reduction_fordward.hpp
  - Rewriting dispatcher routine - caused by changes in the interfaces,
    not logic updates.

### CODE REVIEW

File: sycl/include/sycl/reduction_fwd.hpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
+#pragma once
```
Comment (bader) [inline sycl/include/sycl/reduction_fwd.hpp:1]:
```c++
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```

License header is missing.

**Review by steffenlarsen — APPROVED** [review]
LGTM. Just a minor nit.

File: sycl/include/sycl/handler.hpp [hunk 2]
```diff
@@ -30,6 +30,7 @@
 #include <sycl/nd_item.hpp>
 #include <sycl/nd_range.hpp>
 #include <sycl/property_list.hpp>
+#include <sycl/reduction_fwd.hpp>
```
[the remark this answers is not in the document]
Comment (aelovikov-intel) [reply sycl/include/sycl/handler.hpp]: Done

File: sycl/include/sycl/reduction_fwd.hpp (hunk 1 above)
Comment (aelovikov-intel) [reply sycl/include/sycl/reduction_fwd.hpp]: Done

**Comment by aelovikov-intel** [comment]
Failures:
```
SYCL :: Regression/device_num.cpp
SYCL :: regression/tanh_fix_test.cpp  / on ESIMD
```
seem to be unrelated (e.g., the same are reported on https://github.com/intel/llvm/actions/runs/3430654188/jobs/5718451880). @intel/llvm-gatekeepers can this be merged in?

**Comment by pvchupin** [comment]
@aelovikov-intel, please find the cause of fails. @againull please don't merge until cause is presented and owner identified.
Also there are post-commit fails: https://github.com/intel/llvm/actions/runs/3432868067

**Comment by againull** [comment]
> @aelovikov-intel, please find the cause of fails. @againull please don't merge until cause is presented and owner identified. Also there are post-commit fails: https://github.com/intel/llvm/actions/runs/3432868067

Got it, sorry.

**Comment by aelovikov-intel** [comment]
> Also there are post-commit fails: https://github.com/intel/llvm/actions/runs/3432868067

https://github.com/intel/llvm/pull/7345 should hopefully address those warnings - my local build doesn't have them for some reason.

**Comment by steffenlarsen** [comment]
SYCL :: Regression/device_num.cpp - Reported to owner in https://github.com/intel/llvm-test-suite/pull/1354#issuecomment-1308576318 with fix up in https://github.com/intel/llvm/pull/7336.
SYCL :: regression/tanh_fix_test.cpp - Reported to owner in https://github.com/intel/llvm-test-suite/pull/1361#issuecomment-1308562613 with fix in https://github.com/intel/llvm-test-suite/pull/1372.

**Pull request merged**